tsl/src/compression/recompression.c

```diff
@@ -0,0 +1,1449 @@
+
+	/*
+	 * 1. For every uncompressed row do following:
+	 *   A. If fetched compressed tuple (CR) from (iii) exists:
+	 *       - If uncompressed row matching CR goto step1.
+	 *       - If uncompressed row does not match, save total
+	 *       - fetched uncompressed row count and CR in AffectedSegmentsCxt
+	 *   B. Scan the compressed index
+	 *       i. Fetch the compressed tuple based on segmentby values
+	 *       ii. If matching segmentby tuple is found check the orderby values
+	 *       iii. Save the matching segmentby/orderby compressed row (CR)
+	 *       iv. If segmentby values does not match goto stepB
+	 *       v. If none of compressed rows match the segmentby values mark
+	 *          new_segments_found = true and process all new segmentby uncompressed
+	 *          rows.
+	 *   C. Handle all corner cases as part of steps A,B
+	 */
+	while (tuplesort_gettupleslot(uncompressed_tuplestorecxt->tuplestore,
+								  true /*=forward*/,
+								  false /*=copy*/,
+								  uncompressed_slot,
+								  NULL /*=abbrev*/))
+	{
+		Bitmapset *null_columns = NULL;
+		ScanKeyData *idx_scankey = NULL;
+		int num_idx_scankeys = 0;
+
+		num_uncompressedrows++;
+		slot_getallattrs(uncompressed_slot);
+
+		if (segment_by_keys)
+			pfree(segment_by_keys);
+		if (segment_order_by_keys)
+			pfree(segment_order_by_keys);
+		/* build scan keys to do lookup into uncompressed chunks */
+		segment_by_keys = build_segment_order_by_scankeys(decompressor,
+														  uncompressed_chunk,
+														  uncompressed_slot,
+														  &null_columns,
+														  &num_segkeys,
+														  SEGMENTBY_KEYS);
+		segment_order_by_keys = build_segment_order_by_scankeys(decompressor,
+																uncompressed_chunk,
+																uncompressed_slot,
+																NULL,
+																&num_scankeys,
+																SEGMENTBY_KEYS | ORDERBY_KEYS);
+		if (compressed_tuple_found)
+		{
+			if (null_columns)
+			{
+				/*
+				 * we cannot use HeapKeyTest to check for NULLS, thus perform
+				 * manual checks by calling has_all_nulls()
+				 */
+				uncompressed_tuple = heap_form_tuple(uncompressed_slot->tts_tupleDescriptor,
+													 uncompressed_slot->tts_values,
+													 uncompressed_slot->tts_isnull);
+				if (has_all_nulls(decompressor, compressed_tuple, uncompressed_tuple, null_columns))
+				{
+					/* all nullable segmentby columns have matching NULL values */
+					heap_freetuple(uncompressed_tuple);
+					continue;
+				}
+				heap_freetuple(uncompressed_tuple);
+			}
+			/* reset the flags */
+			segment_exists = false;
+			compressed_tuple_found = false;
+			if (num_segkeys)
+			{
+				HeapKeyTest(compressed_tuple,
+							RelationGetDescr(decompressor->in_rel),
+							num_segkeys,
+							segment_by_keys,
+							segment_exists);
+				HeapKeyTest(compressed_tuple,
+							RelationGetDescr(decompressor->in_rel),
+							num_scankeys,
+							segment_order_by_keys,
+							compressed_tuple_found);
```

Comment:
So, I think what is happening here is that you are testing the uncompressed tuple against the _previously_ found compressed tuple to see if this uncompressed tuple also matches the segment, just like the previously processed uncompressed tuple. 

If this is true, it took me some time to figure this out (and I am still not sure). 

I recommend, instead of having all this code in a _big_ loop, move some of the code into aptly named functions that makes it easier to read and follow the code. For example, here it could make sense to have something like (pseudo code):

```c
HeapTuple compressed_segment_tuple = NULL;

while (uncompressed_tuple = read_next_uncompressed_tuple())
{
    bool tuple_matches_compressed_segment = false;

    while (!tuple_matches_compressed_segment)
    {
        tuple_matches_compressed_segment = tuple_matches_current_compressed_segment(uncompressed_tuple, compressed_segment_tuple)
        if (tuple_matches_compressed_segment) 
        {
              // process uncompressed tuple for matching segment
        }
        else 
        {
            if (compressed_segment_tuple)
                heap_freetuple(compressed_segment_tuple);

            compressed_segment_tuple = read_next_compressed_segment_tuple();
        }
    }
}